tests/acceptance/30_custom_promise_types/24_expireafter.cf

```diff
@@ -0,0 +1,86 @@
+######################################################
+#
+#  Test that expireafter works with custom promise types
+#
+#####################################################
+body common control
+{
+  inputs => { "../default.cf.sub" };
+  bundlesequence  => { default("$(this.promise_filename)") };
+  version => "1.0";
+}
+
+#######################################################
+
+bundle common version_check
+{
+  classes:
+      "python_version_compatible_with_cfengine_library"
+        expression => returnszero(
+          "/usr/bin/python3 -c 'import sys; assert sys.version_info >= (3,6)'",
+          "useshell"
+        );
+
+@if minimum_version(3.20)
+      "custom_promises_supports_expireafter"
+        expression => "any";
+@endif
+}
+
+bundle agent init
+{
+  meta:
+      "test_skip_unsupported"
+        string => "!custom_promises_supports_expireafter|!python_version_compatible_with_cfengine_library";
+
+  files:
+      "$(this.promise_dirname)/cfengine.py"
+        copy_from => local_cp(
+          "$(this.promise_dirname)/../../../modules/promises/cfengine.py"
+        );
+      "/tmp/testfile.txt"
+        delete => tidy;
+}
+
+#######################################################
+
+body action background {
+  background => "true";
+}
+
+bundle agent test
+{
+  meta:
+      "description" -> { "CFE-3435" }
+        string => "Test that expireafter works with custom promise types";
+
+  commands:
+      "$(sys.cf_agent) --inform --file $(this.promise_filename).sub --"
+        action => background;
+      "/bin/sleep 90";
+      "$(sys.cf_agent) --inform --file $(this.promise_filename).sub";
+      "/bin/sleep 100";
```

Comment:
I think you don't need this last sleep.
Since the immediately prior subagent is not executed as background action, the sleep promise should not actuate until after the subagent command has returned.

Reply:
After this sleep finishes we will check the contents of the test output file. This sleep ensures that if the 2nd agent failed at killing the first agent, the first agent (actually the module in it) has time to finish its sleep (which it shouldn't finish if expireafter works correctly) and write the message to the file.

Reply:
^^

Reply:
Actually Nick Anderson is right. By the time the second agent is done – i.e. 90 + 120 sec – the first agent should be finished